### PR TITLE
Flashing Script improvements

### PR DESCRIPTION
**Pull request by mikatammi**
* Fix flashing script to match newest jetpack-nixos
* Big refactor of flashing script, and use somewhat different approach in flash-script generation. This should ensure it is possible to run flash-script on both x86_64-linux and aarch64-linux platforms
* Add version of flash-script that should be possible to run from aarch64-linux platform

### CODE REVIEW

File: modules/hardware/nvidia-jetson-orin/partition-template.nix [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      <allocation_policy> sequential </allocation_policy>
+      <filesystem_type> basic </filesystem_type>
+      <size> 0xFFFFFFFFFFFFFFFF </size>
+      <file_system_attribute> 0 </file_system_attribute>
+      <allocation_attribute> 8 </allocation_attribute>
+      <percent_reserved> 0 </percent_reserved>
+    </partition>
+  '';
+  partitionTemplate = pkgs.runCommand "flash.xml" {} ''
+    head -n 588 ${pkgs.nvidia-jetpack.bspSrc}/bootloader/t186ref/cfg/flash_t234_qspi_sdmmc.xml >$out
```
Comment (brianmcgillion) [inline modules/hardware/nvidia-jetson-orin/partition-template.nix:70]: 588 is a very magic number. Should this be a defined variable with some explanation

File: modules/hardware/nvidia-jetson-orin/partition-template.nix [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    # Replace the section for sdmmc-device with our own section
+    cat ${partitionsEmmc} >>$out
+
+    tail -n 2 ${pkgs.nvidia-jetpack.bspSrc}/bootloader/t186ref/cfg/flash_t234_qspi_sdmmc.xml >>$out
```
Comment (brianmcgillion) [inline modules/hardware/nvidia-jetson-orin/partition-template.nix:75]: not quite as magical but why last 2 lines ?

**Review by brianmcgillion — CHANGES_REQUESTED** [review]

**Comment by vilvo** [comment]
Fails to flash on `aarch64`:
```
[vilvo@blub:~/ghaf]$ git checkout origin/flash_script_fix
Warning: you are leaving 3 commits behind, not connected to
any of your branches:

  0be0329 Add aarch64-linux hosted flash-script
  910118a Update jetpack-nixos in flake.lock
  10ae02f WIP: Fix flashing script to match newest jetpack-nixos

If you want to keep them by creating a new branch, this may be a good time
to do so with:

 git branch <new-branch-name> 0be0329

HEAD is now at 115fd0d Flashing Script improvements

[vilvo@blub:~/ghaf]$ nix build .#packages.aarch64-linux.nvidia-jetson-orin-debug-flash-script
warning: ignoring untrusted flake configuration setting 'extra-trusted-public-keys'.
Pass '--accept-flake-config' to trust it
warning: ignoring untrusted flake configuration setting 'extra-trusted-substituters'.
Pass '--accept-flake-config' to trust it
trace: Obsolete option `hardware.nvidia-jetpack.bootloader.autoUpdate' is used. It was renamed to `hardware.nvidia-jetpack.firmware.autoUpdate'.
[1/6/8 built, 0.0 MiB DL] building nixos-disk-image: [    3.353971] reerror: unexpected end-of-file
(use '--show-trace' to show detailed location information)

[vilvo@blub:~/ghaf]$ nix build --show-trace .#packages.aarch64-linux.nvidia-jetson-orin-debug-flash-script
warning: ignoring untrusted flake configuration setting 'extra-trusted-public-keys'.
Pass '--accept-flake-config' to trust it
warning: ignoring untrusted flake configuration setting 'extra-trusted-substituters'.
Pass '--accept-flake-config' to trust it
trace: Obsolete option `hardware.nvidia-jetpack.bootloader.autoUpdate' is used. It was renamed to `hardware.nvidia-jetpack.firmware.autoUpdate'.
error:
       … while evaluating the attribute 'text' of the derivation 'flash-ghaf'

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/pkgs/stdenv/generic/make-derivation.nix:270:7:

          269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          270|       name =
             |       ^
          271|         let

       … while evaluating call site

       at /nix/store/5wwgw34l43wzgajyh1vhzf4rnnwbkl88-source/modules/hardware/nvidia-jetson-orin/mk-flash-script.nix:25:17:

           24|     ;
           25|   flashScript = import (jetpack-nixos + "/flash-script.nix") {
             |                 ^
           26|     inherit lib;

       … while calling anonymous lambda

       at /nix/store/32lwji0k50gx0nvzjfbiys01k0k165lg-source/flash-script.nix:1:1:

            1| { lib, flash-tools,
             | ^
            2|

       … while evaluating call site

       at /nix/store/32lwji0k50gx0nvzjfbiys01k0k165lg-source/flash-script.nix:42:5:

           41|
           42|   ${lib.optionalString (partitionTemplate != null) "cp ${partitionTemplate} flash.xml"}
             |     ^
           43|   ${lib.optionalString (dtbsDir != null) "cp -r ${dtbsDir}/. kernel/dtb/"}

       … while calling 'optionalString'

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/lib/strings.nix:213:5:

          212|     # String to return if condition is true
          213|     string: if cond then string else "";
             |     ^
          214|

       … while evaluating call site

       at «none»:0: (source not available)

       … while calling 'g'

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/lib/attrsets.nix:417:19:

          416|           g =
          417|             name: value:
             |                   ^
          418|             if isAttrs value && cond value

       … while evaluating call site

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/lib/attrsets.nix:420:20:

          419|               then recurse (path ++ [name]) value
          420|               else f (path ++ [name]) value;
             |                    ^
          421|         in mapAttrs g;

       … while calling anonymous lambda

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/lib/modules.nix:270:72:

          269|           # For definitions that have an associated option
          270|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                                                                        ^
          271|

       … while evaluating the attribute 'value'

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/lib/modules.nix:728:9:

          727|     in warnDeprecation opt //
          728|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          729|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `hardware.nvidia-jetpack.flashScriptOverrides.partitionTemplate':

       … while evaluating the attribute 'mergedValue'

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/lib/modules.nix:763:5:

          762|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          763|     mergedValue =
             |     ^
          764|       if isDefined then

       … while evaluating call site

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/lib/modules.nix:765:12:

          764|       if isDefined then
          765|         if all (def: type.check def.value) defsFinal then type.merge loc defsFinal
             |            ^
          766|         else let allInvalid = filter (def: ! type.check def.value) defsFinal;

       … while calling anonymous lambda

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/lib/modules.nix:765:17:

          764|       if isDefined then
          765|         if all (def: type.check def.value) defsFinal then type.merge loc defsFinal
             |                 ^
          766|         else let allInvalid = filter (def: ! type.check def.value) defsFinal;

       … while evaluating call site

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/lib/modules.nix:765:22:

          764|       if isDefined then
          765|         if all (def: type.check def.value) defsFinal then type.merge loc defsFinal
             |                      ^
          766|         else let allInvalid = filter (def: ! type.check def.value) defsFinal;

       … while calling 'check'

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/lib/types.nix:482:15:

          481|       descriptionClass = "noun";
          482|       check = x: isCoercibleToString x && builtins.substring 0 1 (toString x) == "/";
             |               ^
          483|       merge = mergeEqualOption;

       … while evaluating the attribute 'buildCommand' of the derivation 'flash.xml'

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/pkgs/stdenv/generic/make-derivation.nix:270:7:

          269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          270|       name =
             |       ^
          271|         let

       … while evaluating the attribute 'text' of the derivation 'sdmmc.xml'

       at /nix/store/0havqx505jzm0gq07wdrlzjx94grysdy-source/pkgs/stdenv/generic/make-derivation.nix:270:7:

          269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          270|       name =
             |       ^
          271|         let

       … while realising the context of path '/nix/store/v4mzd40b9sb7rnmgnw5663qw4kd8qkxc-split-images/esp.size'

       at /nix/store/5wwgw34l43wzgajyh1vhzf4rnnwbkl88-source/modules/hardware/nvidia-jetson-orin/partition-template.nix:16:13:

           15|   };
           16|   espSize = builtins.readFile "${images}/esp.size";
             |             ^
           17|   rootSize = builtins.readFile "${images}/root.size";

       error: unexpected end-of-file
 ```

**Review by vilvo — CHANGES_REQUESTED** [review]
Fails to flash on aarch64 with `error: unexpected end-of-file`

**Comment by brianmcgillion** [comment]
❯ nix build .#packages.aarch64-linux.nvidia-jetson-orin-debug-flash-script
trace: Obsolete option `hardware.nvidia-jetpack.bootloader.autoUpdate' is used. It was renamed to `hardware.nvidia-jetpack.firmware.autoUpdate'.

File: modules/hardware/nvidia-jetson-orin/partition-template.nix (hunk 1 above)
Comment (mikatammi) [reply modules/hardware/nvidia-jetson-orin/partition-template.nix:70]: This is only moved from the previous place but of course I can add some comments

**Comment by mikatammi** [comment]
> trace: Obsolete option hardware.nvidia-jetpack.bootloader.autoUpdate' is used. It was renamed to `hardware.nvidia-jetpack.firmware.autoUpdate'.

This is probably something that is internal to jetpack-nixos. There is no reference to nvidia-jetpack.bootloader anywhere in ghaf

**Comment by mikatammi** [comment]
> Fails to flash on aarch64 with `error: unexpected end-of-file`

This probably an issue in your setup. Have you tried to build just the basic image from `.#packages.aarch64-linux.nvidia-jetson-orin-debug` and seen if that works? From the log I can see that it starts to run QEMU for the `nixos-disk-image`but it looks like it fails before it gets to completion. I just tried to do the build of `.#packages.aarch64-linux.nvidia-jetson-orin-debug-flash-script` on fresh install on ghaf with no issues.

**Comment by mikatammi** [comment]
There is actually one real issue that needs to be fixed and that's why I changed this back to draft. The esp.img does not get copied or symlinked into place before flashing.

**Comment by vilvo** [comment]
> > Fails to flash on aarch64 with `error: unexpected end-of-file`
> 
> This probably an issue in your setup. Have you tried to build just the basic image from `.#packages.aarch64-linux.nvidia-jetson-orin-debug` and seen if that works? 

It works just fine:
```
[vilvo@blub:~/ghaf]$ nix build .#packages.aarch64-linux.nvidia-jetson-orin-debug
warning: ignoring untrusted flake configuration setting 'extra-trusted-public-keys'.
Pass '--accept-flake-config' to trust it
warning: ignoring untrusted flake configuration setting 'extra-trusted-substituters'.
Pass '--accept-flake-config' to trust it
trace: Obsolete option `hardware.nvidia-jetpack.bootloader.autoUpdate' is used. It was renamed to `hardware.nvidia-jetpack.firmware.autoUpdate'.

[vilvo@blub:~/ghaf]$ echo $?
0
[vilvo@blub:~/ghaf]$ ls -la result
lrwxrwxrwx 1 vilvo users 60 May 23 14:01 result -> /nix/store/gvrls35hxc3k6d6b0734kadg7kk8m4am-nixos-disk-image
```

> From the log I can see that it starts to run QEMU for the `nixos-disk-image`but it looks like it fails before it gets to completion. 

Indeed, it start processing the `nixos-disk-image` after the image build on my NixOS setup (Apple M1) but the flash script part fails consistently every time. Happy to help debugging if you get ideas.

> I just tried to do the build of `.#packages.aarch64-linux.nvidia-jetson-orin-debug-flash-script` on fresh install on ghaf with no issues.

Good on you.

File: modules/hardware/nvidia-jetson-orin/partition-template.nix (hunk 1 above)
Comment (mikatammi) [reply modules/hardware/nvidia-jetson-orin/partition-template.nix]: There's now constants set in the let expression above, and some commentary about the magic

**Comment by vilvo** [comment]
Interesting enough - I did some more testing as root (earlier as user) and found my nixos dotfiles had `git` installed only for user, not system-wide. So I moved `git` to
```
  environment.systemPackages = with pkgs;
    [
      git
    ]
```
and sure enough - I ran and the target created some empty git repositories part of process and finished.
```
[root@blub:/home/vilvo/ghaf]# nix run .#packages.aarch64-linux.nvidia-jetson-orin-debug-flash-script
###############################################################################
# L4T BSP Information:
# R35 , REVISION: 3.1
###############################################################################
grep: (standard input): binary file matches
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
Error: probing the target board failed.
       Make sure the target board is connected through
       USB port and is in recovery mode.
[root@blub:/home/vilvo/ghaf]# dmesg
...
[21718.581413] usb 1-1.1: USB disconnect, device number 8
[21719.327364] usb 1-1.1: new high-speed USB device number 9 using xhci-hcd
[21719.466402] usb 1-1.1: New USB device found, idVendor=0955, idProduct=7023, bcdDevice= 4.01
[21719.466414] usb 1-1.1: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[21719.466420] usb 1-1.1: Product: APX
[21719.466425] usb 1-1.1: Manufacturer: NVIDIA Corp.       
```

**Comment by brianmcgillion** [comment]
It builds well now. I have the same problem that the device does not get picked up as in recovery mode. Though I previously had the same issue.

**Review by brianmcgillion — APPROVED** [review]

**Comment by mikatammi** [comment]
I added comments to the source code, that the aarch64-linux flashing script is experimental and might not work, but still left it in place

**Pull request merged**